### PR TITLE
[OM-92009]: add flag to enable pod move across cluster

### PR DESCRIPTION
**Pull request by libai98**
code for adding a flag option to enable pod move across cluster.

### CODE REVIEW

File: pkg/discovery/dtofactory/cluster_dto_builder.go [hunk 1]
```diff
@@ -14,29 +14,37 @@ import (
...
 	}
 }
 
 // GetClusterKey constructs the commodity key sold by the cluster entity, by adding a prefix to the cluster id.
 // Use of a prefix is to distinguish the same key used by the node entities in the cluster
-func GetClusterKey(clusterId string) string {
+func GetClusterKey(clusterId string, enablePodClusterMoves bool) string {
+	if enablePodClusterMoves {
+		return ClusterFreeKey
+	}
```
Comment (chlam4) [inline pkg/discovery/dtofactory/cluster_dto_builder.go:47]: I'm not sure if we should allow customization of the key between cluster and namespace entities.  Let's discuss.  I think we want such a customization between pods and nodes/VMs (so pods can move across VMs in different clusters).

File: cmd/kubeturbo/app/kubeturbo_builder.go [hunk 2]
```diff
@@ -134,6 +135,10 @@ type VMTServer struct {
 	// The Openshift SCC list allowed for action execution
 	sccSupport []string
 
+	// Allow pod move across cluster
+	// The default is false
+	enablePodClusterMoves bool
```
[the remark this answers is not in the document]
Comment (libai98) [reply cmd/kubeturbo/app/kubeturbo_builder.go:140]: That is a good suggestion! We can discuss on this in the standup meeting.

[the remark this answers is not in the document]
Comment (libai98) [reply cmd/kubeturbo/app/kubeturbo_builder.go]: code modified based on our discussion.

File: pkg/discovery/dtofactory/cluster_dto_builder.go (hunk 1 above)
Comment (libai98) [reply pkg/discovery/dtofactory/cluster_dto_builder.go]: Code modified based on our discussion. Thanks!

File: pkg/discovery/dtofactory/node_entity_dto_builder.go [hunk 3]
```diff
@@ -221,7 +228,7 @@ func (builder *nodeEntityDTOBuilder) getNodeCommoditiesSold(node *api.Node, clus
 		if isResizeable, exists := resizableCommodities[commSold.GetCommodityType()]; exists {
 			commSold.Resizable = &isResizeable
 		}
-	}
```
Comment (chlam4) [inline pkg/discovery/dtofactory/node_entity_dto_builder.go:224]: Accidentally added?

File: pkg/discovery/dtofactory/node_entity_dto_builder.go [hunk 4]
```diff
@@ -239,8 +246,16 @@ func (builder *nodeEntityDTOBuilder) getNodeCommoditiesSold(node *api.Node, clus
 	}
 
 	// Cluster commodity.
+	var clusterCommKey string
+	if len(strings.TrimSpace(builder.clusterKeyInjected)) != 0 {
+		clusterCommKey = builder.clusterKeyInjected
+		glog.V(5).Infof("Injecting cluster key for Node %s with key : %s", node.Name, clusterCommKey)
```
Comment (chlam4) [inline pkg/discovery/dtofactory/node_entity_dto_builder.go:252]: I think @ading1977 said we don't have V5.  Maybe changing it to V4?

File: pkg/discovery/dtofactory/node_entity_dto_builder.go [hunk 5]
```diff
@@ -239,8 +246,16 @@ func (builder *nodeEntityDTOBuilder) getNodeCommoditiesSold(node *api.Node, clus
...
+	} else {
+		clusterCommKey = clusterId
+		glog.V(5).Infof("Adding cluster key for Node %s with key : %s", node.Name, clusterCommKey)
```
Comment (chlam4) [inline pkg/discovery/dtofactory/node_entity_dto_builder.go:255]: Similar comment on the V5.

File: pkg/discovery/dtofactory/pod_entity_dto_builder.go [hunk 6]
```diff
@@ -423,7 +430,16 @@ func (builder *podEntityDTOBuilder) getPodCommoditiesBought(
 	if err != nil {
 		glog.Errorf("Failed to get %s used for current Kubernetes Cluster %s", metrics.Cluster, clusterInfo)
 	} else {
-		clusterCommodityKey, ok := clusterInfo.GetValue().(string)
+		var clusterCommodityKey string
+		var ok bool
+		if len(strings.TrimSpace(builder.clusterKeyInjected)) != 0 {
```
Comment (chlam4) [inline pkg/discovery/dtofactory/pod_entity_dto_builder.go:435]: This is the only place using `TrimSpace`.  We should probably make it consistent across all places.  Perhaps we do the trimming at the very top and propagate only the trimmed version down to every place?

**Review by chlam4 — COMMENTED** [review]
Thanks @libai98!  Looks good to me other than a few nit comments.  Shall we test it out before merging?

File: pkg/discovery/dtofactory/node_entity_dto_builder.go (hunk 3 above)
Comment (libai98) [reply pkg/discovery/dtofactory/node_entity_dto_builder.go]: Yes, fixed.  :)  Thanks!

File: pkg/discovery/dtofactory/pod_entity_dto_builder.go (hunk 6 above)
Comment (libai98) [reply pkg/discovery/dtofactory/pod_entity_dto_builder.go:435]: There is a long chain for the injected cluster key to be passed down from the command arg to pod and node build.
Even I do not assigned a value, the default empty string will be passed along. I am trying not to do the checking and trimming for empty string until this key is used for creating cluster access commodity in pod and node DTO builder.

**Comment by libai98** [comment]
> Thanks @libai98! Looks good to me other than a few nit comments. Shall we test it out before merging?

Yes, I will try to test that out before merging. Thanks for all the help on this!

File: pkg/discovery/dtofactory/node_entity_dto_builder.go (hunk 4 above)
Comment (libai98) [reply pkg/discovery/dtofactory/node_entity_dto_builder.go]: Yes.

File: pkg/discovery/dtofactory/node_entity_dto_builder.go (hunk 5 above)
Comment (libai98) [reply pkg/discovery/dtofactory/node_entity_dto_builder.go]: Yes. Thanks for the quick review!

**Review by chlam4 — APPROVED** [review]

**Pull request merged**